packages/fuel-indexer-schema/src/db/graphql.rs

```diff
@@ -261,43 +261,136 @@ impl Operation {
+                    &mut selections
+                        .get_selections()
+                        .drain(..)
+                        .rev()
+                        .collect::<Vec<Selection>>(),
+                );
+
+                let mut last_seen_entities_len = entities.len();
+
+                while let Some(current) = queue.pop() {
+                    let entity_name = entities.pop().unwrap();
+
+                    // If a selection was processed without adding additional selections
+                    // to the queue, then check the entity of the selection against the
+                    // current nesting level. If they differ, then the operation has moved
+                    // out of a child entity into a parent entity.
+                    if let Some(current_nesting_level) = nested_entity_stack.last() {
+                        if entities.len() < last_seen_entities_len
+                            && current_nesting_level != &entity_name
+                        {
+                            let _ = nested_entity_stack.pop();
+                            elements.push(QueryElement::ObjectClosingBoundary);
                         }
-                    })
-                    .collect();
+                    }
 
-                let column_text = columns.join(", ");
+                    last_seen_entities_len = entities.len();
 
-                let namespace = format!("{namespace}_{identifier}");
-                let mut query = format!(
-                    "SELECT {} FROM {}.{}",
-                    column_text,
-                    namespace,
-                    name.to_lowercase()
-                );
+                    if let Selection::Field(field_name, _f, subselections) = current {
+                        if subselections.selections.is_empty() {
+                            elements.push(QueryElement::Field {
+                                key: field_name.clone(),
+                                value: format!(
+                                    "{namespace}_{identifier}.{entity_name}.{field_name}"
+                                ),
+                            });
+                        } else {
+                            // If the current entity has a foreign key on the current
+                            // selection, join the foreign table on that primary key
+                            // and set the field as the innermost entity by pushing to the stack.
+                            if let Some(field_to_foreign_key) =
+                                schema.foreign_keys.get(&entity_name.to_lowercase())
+                            {
+                                if let Some(foreign_key) =
+                                    field_to_foreign_key.get(&field_name.to_lowercase())
+                                {
+                                    let reference_table =
+                                        format!("{namespace}_{identifier}.{field_name}");
```

Comment:
Maybe we can make some of this stuff a bit more self-documenting?
```suggestion
                                        fully_qualitified_field_name(&namespace, &identifier, &field_name);
```
Where
```rust
fn fully_qualitified_field_name(namespace: &str, identifier: &str, field_name: &str) -> String {
    format!("{namespace}_{identifier}.{field_name}")
}